src/pymmcore_plus/mda/_engine.py

```diff
@@ -59,6 +115,18 @@ def exec_event(self, event: MDAEvent) -> Any:
         self._mmc.snapImage()
         return EventPayload(image=self._mmc.getImage())
 
+    def _execute_autofocus(self, autofocus_event: AutoFocusParams) -> float:
+        """Perform the autofocus.
+
+        Returns the correction to be applied to the focus motor position.
+        """
+        self._mmc.setPosition(
+            autofocus_event.autofocus_z_device_name,
+            cast(float, autofocus_event.af_motor_offset),
+        )
+        self._mmc.fullFocus()
+        return self._mmc.getZPosition() - cast(float, autofocus_event.z_stage_position)
```

Comment:
I've found that there absolutely need to be some waits in here. either `waitForSystem` or `waitForDevices`.

On some of my older systems i've also found it be significantly more robust to catch an error in autofocus and retry once or twice. Anecdotally on the olympus system I use 1 retry works most of the time, and 2 retries has yet to fail even over a 16 hour timelapse.

Here's the equivalent section of my engine:

https://github.com/ianhi/raman-mda-engine/blob/f7238bc286aa9642331eb5c6215c458e31327115/raman_mda_engine/_engine.py#L273-L284

Reply:
> I've found that there absolutely need to be some waits in here.

without the waits on some systems the `getZPosition` may not report on the correctly focused stage position.